README.md




### Installation

```bash
npm i @fanitrade/fani-solana-tokenlist
```

```bash
yarn add @fanitrade/fani-solana-tokenlist
```

### Examples

#### Query available tokens

```typescript
new TokenListProvider().resolve().then((tokens) => {
  const tokenList = tokens.filterByClusterSlug('mainnet-beta').getList();
  console.log(tokenList);
});
```


### Query available tokens on chain
```typescript
import { getTokenMetaData } from "@fanitrade/fani-solana-tokenlist";
import { Connection, clusterApiUrl } from "@solana/web3.js";

const connection = new Connection(clusterApiUrl('devnet'));

( async () => {
  const mint = "7SLdnLCBDFEBAZvzdh9vGeuoZBD165ajKfAxSgyjFjwv";

  const metadata = await getTokenMetaData(connection , mint);
  console.log(metadata);

})()
```

#### Response
```json
{
  "name": "Token NAME",
  "address": "7SLdnLCBDFEBAZvzdh9vGeuoZBD165ajKfAxSgyjFjwv",
  "symbol": "test",
  "image": "https://images.google.com/images/branding/googlelogo/1x/googlelogo_light_color_272x92dp.png",
  "extensions": {
    "website": "",
    "facebook": "",
    "twitter": "",
    "github": "",
    "discord": "",
    "telegram": ""
  }
}
```


### Adding new token
```typescript
import { addTokenToList } from "@fanitrade/fani-solana-tokenlist";


( async () => {
      const token =  await addTokenToList("YOUR_GITHUB_ACCESS_TOKEN" , {
        address : "MINT ADDRESS",
        symbol: "SYL",
         name: "TOKEN NAME",
        decimals:"0",
        logoURI:"LOG0_URL.png", 
        extensions: {
            website: "WEBSITE_URL",
            facebook: "FACEBOOK_URL",
            twitter: "TWITTER_URL",
        }
    })
    console.log(token)
})()
```

#### Response 
```javascript
{
  status: "Created",
  message: "Token Created"
}
```

```javascript
 {
  status: 'Failed',
  message: 'Error Creating Token',
  detail: 'Reference Already exist or invalid input'
}
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.3.19](https://github.com/FaniTrade-Investment-L-L-C/fani-token-list/compare/v1.3.17...v1.3.19) (2022-07-01)

### [1.3.17](https://github.com/FaniTrade-Investment-L-L-C/fani-token-list/compare/v1.3.15...v1.3.17) (2022-07-01)

### [1.3.15](https://github.com/FaniTrade-Investment-L-L-C/fani-token-list/compare/v1.3.13...v1.3.15) (2022-06-30)

### [1.3.13](https://github.com/FaniTrade-Investment-L-L-C/fani-token-list/compare/v1.3.11...v1.3.13) (2022-06-30)

### [1.3.11](https://github.com/FaniTrade-Investment-L-L-C/fani-token-list/compare/v1.3.9...v1.3.11) (2022-06-29)

### [1.3.9](https://github.com/FaniTrade-Investment-L-L-C/fani-token-list/compare/v1.3.7...v1.3.9) (2022-06-28)

### [1.3.7](https://github.com/FaniTrade-Investment-L-L-C/fani-token-list/compare/v1.3.5...v1.3.7) (2022-06-28)

### [1.3.5](https://github.com/FaniTrade-Investment-L-L-C/fani-token-list/compare/v1.3.3...v1.3.5) (2022-06-28)

### [1.3.3](https://github.com/FaniTrade-Investment-L-L-C/fani-token-list/compare/v1.3.1...v1.3.3) (2022-06-28)

### [1.3.1](https://github.com/FaniTrade-Investment-L-L-C/fani-token-list/compare/v1.2.0...v1.3.1) (2022-06-28)

## [1.2.0](https://github.com/FaniTrade-Investment-L-L-C/fani-token-list/compare/v0.2.4574...v1.2.0) (2022-06-28)


### Features

* add deprecated information ([eeafc4d](https://github.com/FaniTrade-Investment-L-L-C/fani-token-list/commit/eeafc4d52d8ce3db61e64cf338df479cb55587f7))

### 0.2.4574 (2022-06-21)

### 0.2.4573 (2022-06-20)

### 0.2.4572 (2022-06-20)

### 0.2.4571 (2022-06-20)

### 0.2.4570 (2022-06-20)

### 0.2.4569 (2022-06-20)

### 0.2.4568 (2022-06-20)

### 0.2.4567 (2022-06-20)

### 0.2.4566 (2022-06-20)

### 0.2.4565 (2022-06-20)

### 0.2.4564 (2022-06-20)

### 0.2.4563 (2022-06-20)

### 0.2.4562 (2022-06-20)

### 0.2.4561 (2022-06-20)

### 0.2.4560 (2022-06-20)

### 0.2.4559 (2022-06-20)

### 0.2.4558 (2022-06-20)

### 0.2.4557 (2022-06-20)

### 0.2.4556 (2022-06-19)

### 0.2.4555 (2022-06-19)

### 0.2.4554 (2022-06-19)

### 0.2.4553 (2022-06-19)

### 0.2.4552 (2022-06-19)

### 0.2.4551 (2022-06-19)

### 0.2.4550 (2022-06-19)

### 0.2.4549 (2022-06-19)

### 0.2.4548 (2022-06-19)

### 0.2.4547 (2022-06-19)

### 0.2.4546 (2022-06-19)

### 0.2.4545 (2022-06-19)

### 0.2.4544 (2022-06-19)

### 0.2.4543 (2022-06-19)

### 0.2.4542 (2022-06-19)

### 0.2.4541 (2022-06-19)

### 0.2.4540 (2022-06-19)

### 0.2.4539 (2022-06-19)

### 0.2.4538 (2022-06-18)

### 0.2.4537 (2022-06-18)

### 0.2.4536 (2022-06-18)

### 0.2.4535 (2022-06-18)

### 0.2.4534 (2022-06-18)

### 0.2.4533 (2022-06-18)

### 0.2.4532 (2022-06-18)

### 0.2.4531 (2022-06-18)

### 0.2.4530 (2022-06-18)

### 0.2.4529 (2022-06-17)

### 0.2.4528 (2022-06-17)

### 0.2.4527 (2022-06-17)

### 0.2.4526 (2022-06-17)

### 0.2.4525 (2022-06-17)

### 0.2.4524 (2022-06-17)

### 0.2.4523 (2022-06-17)

### 0.2.4522 (2022-06-17)

### 0.2.4521 (2022-06-17)

### 0.2.4520 (2022-06-17)

### 0.2.4519 (2022-06-17)

### 0.2.4518 (2022-06-17)

### 0.2.4517 (2022-06-17)

### 0.2.4516 (2022-06-17)

### 0.2.4515 (2022-06-17)

### 0.2.4514 (2022-06-16)

### 0.2.4513 (2022-06-16)

### 0.2.4512 (2022-06-16)

### 0.2.4511 (2022-06-16)

### 0.2.4510 (2022-06-16)

### 0.2.4509 (2022-06-16)

### 0.2.4508 (2022-06-16)

### 0.2.4507 (2022-06-16)

### 0.2.4506 (2022-06-16)

### 0.2.4505 (2022-06-16)

### 0.2.4504 (2022-06-16)

### 0.2.4503 (2022-06-16)

### 0.2.4502 (2022-06-16)

### 0.2.4501 (2022-06-16)

### 0.2.4500 (2022-06-16)

### 0.2.4499 (2022-06-16)

### 0.2.4498 (2022-06-16)

### 0.2.4497 (2022-06-16)

### 0.2.4496 (2022-06-16)

### 0.2.4495 (2022-06-16)

### 0.2.4494 (2022-06-15)

### 0.2.4493 (2022-06-15)

### 0.2.4492 (2022-06-15)

### 0.2.4491 (2022-06-15)

### 0.2.4490 (2022-06-15)

### 0.2.4489 (2022-06-15)

### 0.2.4488 (2022-06-15)

### 0.2.4487 (2022-06-15)

### 0.2.4486 (2022-06-15)

### 0.2.4485 (2022-06-15)

### 0.2.4484 (2022-06-15)

### 0.2.4483 (2022-06-15)

### 0.2.4482 (2022-06-15)

### 0.2.4481 (2022-06-15)

### 0.2.4480 (2022-06-15)

### 0.2.4479 (2022-06-15)

### 0.2.4478 (2022-06-15)

### 0.2.4477 (2022-06-15)

### 0.2.4476 (2022-06-15)

### 0.2.4475 (2022-06-15)

### 0.2.4474 (2022-06-14)

### 0.2.4473 (2022-06-14)

### 0.2.4472 (2022-06-14)

### 0.2.4471 (2022-06-14)

### 0.2.4470 (2022-06-14)

### 0.2.4469 (2022-06-14)

### 0.2.4468 (2022-06-14)

### 0.2.4467 (2022-06-14)

### 0.2.4466 (2022-06-14)

### 0.2.4465 (2022-06-14)

### 0.2.4464 (2022-06-14)

### 0.2.4463 (2022-06-14)

### 0.2.4462 (2022-06-14)

### 0.2.4461 (2022-06-14)

### 0.2.4460 (2022-06-14)

### 0.2.4459 (2022-06-14)

### 0.2.4458 (2022-06-14)

### 0.2.4457 (2022-06-14)

### 0.2.4456 (2022-06-14)

### 0.2.4455 (2022-06-14)

### 0.2.4454 (2022-06-13)

### 0.2.4453 (2022-06-13)

### 0.2.4452 (2022-06-13)

### 0.2.4451 (2022-06-13)

### 0.2.4450 (2022-06-13)

### 0.2.4449 (2022-06-13)

### 0.2.4448 (2022-06-13)

### 0.2.4447 (2022-06-13)

### 0.2.4446 (2022-06-13)

### 0.2.4445 (2022-06-13)

### 0.2.4444 (2022-06-13)

### 0.2.4443 (2022-06-13)

### 0.2.4442 (2022-06-13)

### 0.2.4441 (2022-06-13)

### 0.2.4440 (2022-06-13)

### 0.2.4439 (2022-06-13)

### 0.2.4438 (2022-06-13)

### 0.2.4437 (2022-06-12)

### 0.2.4436 (2022-06-12)

### 0.2.4435 (2022-06-12)

### 0.2.4434 (2022-06-12)

### 0.2.4433 (2022-06-12)

### 0.2.4432 (2022-06-12)

### 0.2.4431 (2022-06-12)

### 0.2.4430 (2022-06-12)

### 0.2.4429 (2022-06-12)

### 0.2.4428 (2022-06-12)

### 0.2.4427 (2022-06-12)

### 0.2.4426 (2022-06-12)

### 0.2.4425 (2022-06-12)

### 0.2.4424 (2022-06-12)

### 0.2.4423 (2022-06-12)

### 0.2.4422 (2022-06-12)

### 0.2.4421 (2022-06-11)

### 0.2.4420 (2022-06-11)

### 0.2.4419 (2022-06-11)

### 0.2.4418 (2022-06-11)

### 0.2.4417 (2022-06-11)

### 0.2.4416 (2022-06-11)

### 0.2.4415 (2022-06-11)

### 0.2.4414 (2022-06-11)

### 0.2.4413 (2022-06-11)


src/utils/id.ts
import { PublicKey } from "@solana/web3.js";

export const WRAPPED_SOL_MINT = new PublicKey(
  "So11111111111111111111111111111111111111112"
);
export const TOKEN_PROGRAM_ID = new PublicKey(
  "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
);

export const SPL_ASSOCIATED_TOKEN_ACCOUNT_PROGRAM_ID = new PublicKey(
  "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL"
);
export const BPF_UPGRADE_LOADER_ID = new PublicKey(
  "BPFLoaderUpgradeab1e11111111111111111111111"
);

export const METADATA_PROGRAM_ID = new PublicKey(
  "metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s"
  // 'GCUQ7oWCzgtRKnHnuJGxpr5XVeEkxYUXwTKYcqGtxLv4',
);

export const MEMO_ID = new PublicKey(
  "MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr"
);

export const VAULT_ID = new PublicKey(
  "vau1zxA2LbssAUEF7Gpw91zMM1LvXrvpzJtmZ58rPsn"
  // '41cCnZ1Z1upJdtsS1tzFGR34cPFgJLzvJFmgYKpCqkz7',
);

export const AUCTION_ID = new PublicKey(
  "auctxRXPeJoc4817jDhf4HbjnhEcr1cCXenosMhK5R8"
  // '6u5XVthCStUfmNrYhFsST94oKxzwEZfZFHFhiCnB2nR1',
);

export const METAPLEX_ID = new PublicKey(
  "p1exdMJcjVao65QdewkaZRUnU6VPSXhus9n2GzWfh98"
  // '98jcGaKLKx9vv33H9edLUXAydrSipHhJGDQuPXBVPVGp',
);

export const SYSTEM = new PublicKey("11111111111111111111111111111111");

export const ENABLE_FEES_INPUT = false;

// legacy pools are used to show users contributions in those pools to allow for withdrawals of funds
export const PROGRAM_IDS = [
  {
    name: "mainnet-beta",
  },
  {
    name: "testnet",
  },

  {
    name: "devnet",
  },
  {
    name: "localnet",
  },
];

export const programIds = () => {
  return {
    token: TOKEN_PROGRAM_ID,
    associatedToken: SPL_ASSOCIATED_TOKEN_ACCOUNT_PROGRAM_ID,
    bpf_upgrade_loader: BPF_UPGRADE_LOADER_ID,
    system: SYSTEM,
    metadata: METADATA_PROGRAM_ID,
    memo: MEMO_ID,
    vault: VAULT_ID,
    auction: AUCTION_ID,
    metaplex: METAPLEX_ID,
  };
};

src/lib/tokenlist.ts
import * as anchor from '@project-serum/anchor';
import { Connection, PublicKey } from '@solana/web3.js';
import axios from 'axios';
import { fetch } from 'cross-fetch';
import Joi from 'joi';

import { decodeMetadata } from '../utils/Metadata';
import { sendToSolanaTokenList } from '../utils/sendTokenToSolanaTokenList';
import tokenlist from './../tokens/solana.tokenlist.json';
const TOKEN_METADATA_PROGRAM_ID = new anchor.web3.PublicKey(
  'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s'
);


const schema = Joi.object({
  chainId: Joi.number().
               required(),
  address: Joi.string()
      .required(),

  symbol: Joi.string()
             .required(),

  name: Joi.string()
           .required(),

  decimals: Joi.string().
           required(),
  logoURI: Joi.string()
              .required(),
  extensions: Joi.object().optional()
})



export enum ENV {
  MainnetBeta = 101,
  Testnet = 102,
  Devnet = 103
}

export interface TokenList {
  readonly name: string;
  readonly logoURI: string;
  readonly tags: { [tag: string]: TagDetails };
  readonly timestamp: string;
  readonly tokens: TokenInfo[];
}

export interface TagDetails {
  readonly name: string;
  readonly description: string;
}

export interface TokenExtensions {
  readonly website?: string;
  readonly bridgeContract?: string;
  readonly assetContract?: string;
  readonly address?: string;
  readonly explorer?: string;
  readonly twitter?: string;
  readonly github?: string;
  readonly medium?: string;
  readonly tgann?: string;
  readonly tggroup?: string;
  readonly discord?: string;
  readonly serumV3Usdt?: string;
  readonly serumV3Usdc?: string;
  readonly coingeckoId?: string;
  readonly imageUrl?: string;
  readonly description?: string;
}

export interface TokenInfo {
  readonly chainId: number;
  readonly address: string;
  readonly name: string;
  readonly decimals: number;
  readonly symbol: string;
  readonly logoURI?: string;
  readonly tags?: string[];
  readonly extensions?: TokenExtensions;
}

export type TokenInfoMap = Map<string, TokenInfo>;

export const CLUSTER_SLUGS: { [id: string]: ENV } = {
  'mainnet-beta': ENV.MainnetBeta,
  testnet: ENV.Testnet,
  devnet: ENV.Devnet
};

export class GitHubTokenListResolutionStrategy {
  repositories = [
    'https://raw.githubusercontent.com/FaniTrade-Investment-L-L-C/fani-token-list/main/src/tokens/solana.tokenlist.json'
  ];

  resolve = () => {
    return queryJsonFiles(this.repositories);
  };
}

export class CDNTokenListResolutionStrategy {
  repositories = [
    'https://cdn.jsdelivr.net/gh/solana-labs/token-list@latest/src/tokens/solana.tokenlist.json'
  ];

  resolve = () => {
    return queryJsonFiles(this.repositories);
  };
}

export class SolanaTokenListResolutionStrategy {
  repositories = ['https://token-list.solana.com/solana.tokenlist.json'];

  resolve = () => {
    return queryJsonFiles(this.repositories);
  };
}

const queryJsonFiles = async (files: string[]) => {
  const responses: TokenList[] = (await Promise.all(
    files.map(async (repo) => {
      try {
        const response = await fetch(repo);
        const json = (await response.json()) as TokenList;
        return json;
      } catch {
        console.info(
          `@solana/token-registry: falling back to static repository.`
        );
        return tokenlist;
      }
    })
  )) as TokenList[];

  return responses
    .map((tokenlist: TokenList) => tokenlist.tokens || [])
    .reduce((acc, arr) => (acc as TokenInfo[]).concat(arr), []);
};

export enum Strategy {
  GitHub = 'GitHub',
  Static = 'Static',
  Solana = 'Solana',
  CDN = 'CDN'
}

export class StaticTokenListResolutionStrategy {
  resolve = () => {
    return tokenlist.tokens || [];
  };
}

export class TokenListProvider {
  static strategies = {
    [Strategy.GitHub]: new GitHubTokenListResolutionStrategy(),
    [Strategy.Static]: new StaticTokenListResolutionStrategy(),
    [Strategy.Solana]: new SolanaTokenListResolutionStrategy(),
    [Strategy.CDN]: new CDNTokenListResolutionStrategy()
  };

  resolve = async (
    strategy: Strategy = Strategy.CDN
  ): Promise<TokenListContainer> => {
    return new TokenListContainer(
      await TokenListProvider.strategies[strategy].resolve()
    );
  };
}

async function getMetadata(mint: PublicKey): Promise<PublicKey> {
  return (
    await anchor.web3.PublicKey.findProgramAddress(
      [
        Buffer.from('metadata'),
        TOKEN_METADATA_PROGRAM_ID.toBuffer(),
        mint.toBuffer()
      ],
      TOKEN_METADATA_PROGRAM_ID
    )
  )[0];
}


export async function addTokenToList(
  gitAccessToken: string,
  tokenDetails: any
) {
  // const { address, symbol, name, decimals, logoURI, extensions } = tokenDetails;
 
  console.log(gitAccessToken);
  const { error } = schema.validate({ ...tokenDetails });
  if(error) {
    console.log(error)
    return({
      status: "Failed",
      message: "Error Creating Token",
      detail: error.details[0].message
    })
  }
  else {
    const isEmpty = Object.values(tokenDetails.extensions).every(x => x === null || x === '');
   if(isEmpty) {
     delete tokenDetails.extensions;
     console.log("deleting extension obj  values...");
     console.log(tokenDetails)
   } else {
     Object.keys(tokenDetails.extensions).forEach((k) => tokenDetails.extensions[k] == '' && delete tokenDetails.extensions[k]);
      console.log("deleting unused values");
      console.log(tokenDetails);
   }

   const reuslt = await sendToSolanaTokenList(
    gitAccessToken,
    tokenDetails
  );

  return reuslt
  }
}

export async function getTokenMetaData(
  connection: Connection,
  mint: string
): Promise<any> {
  let fetchedData: any;
  const metaData = await getMetadata(new PublicKey(mint));
  const accountInfo = await connection.getParsedAccountInfo(metaData);
  const decodedData = decodeMetadata(accountInfo?.value?.data);
  if (decodedData) {
    const url = encodeURI(decodedData.data.uri);
    return axios.get(url.split('%00%00%00%00')[0]).then((res) => {
      fetchedData = res.data;
      return new Promise(function (resolve) {
        resolve(fetchedData);
      });
    });
  } else {
    return {};
    // // @ts-ignore
    // new Promise(function (resolve, reject) {
    //   reject({});
    // });
  }
}
export class TokenListContainer {
  constructor(private tokenList: TokenInfo[]) {}

  filterByTag = (tag: string) => {
    return new TokenListContainer(
      this.tokenList.filter((item) => (item.tags || []).includes(tag))
    );
  };

  filterByChainId = (chainId: number | ENV) => {
    return new TokenListContainer(
      this.tokenList.filter((item) => item.chainId === chainId)
    );
  };

  excludeByChainId = (chainId: number | ENV) => {
    return new TokenListContainer(
      this.tokenList.filter((item) => item.chainId !== chainId)
    );
  };

  excludeByTag = (tag: string) => {
    return new TokenListContainer(
      this.tokenList.filter((item) => !(item.tags || []).includes(tag))
    );
  };

  filterByClusterSlug = (slug: string) => {
    if (slug in CLUSTER_SLUGS) {
      return this.filterByChainId(CLUSTER_SLUGS[slug]);
    }
    throw new Error(
      `Unknown slug: ${slug}, please use one of ${Object.keys(CLUSTER_SLUGS)}`
    );
  };

  getList = () => {
    return this.tokenList;
  };
}


src/utils/Metadata.ts
import { PublicKey } from '@solana/web3.js';
import { BinaryReader, BinaryWriter, deserializeUnchecked } from 'borsh';
import base58 from 'bs58';

export const extendBorsh = () => {
  (BinaryReader.prototype as any).readPubkey = function () {
    const reader = (this as unknown) as BinaryReader;
    const array = reader.readFixedArray(32);
    return new PublicKey(array);
  };

  (BinaryWriter.prototype as any).writePubkey = function (value: PublicKey) {
    const writer = (this as unknown) as BinaryWriter;
    writer.writeFixedArray(value.toBuffer());
  };

  (BinaryReader.prototype as any).readPubkeyAsString = function () {
    const reader = (this as unknown) as BinaryReader;
    const array = reader.readFixedArray(32);
    return base58.encode(array);
  };

  (BinaryWriter.prototype as any).writePubkeyAsString = function (
    value: any
  ): void {
    const writer = (this as unknown) as BinaryWriter;
    writer.writeFixedArray(base58.decode(value));
  };
};

extendBorsh();

const MetadataKey = {
  Uninitialized: 0,
  MetadataV1: 4,
  EditionV1: 1,
  MasterEditionV1: 2,
  MasterEditionV2: 6,
  EditionMarker: 7
};

export const METADATA_PREFIX = 'metadata';
export const EDITION = 'edition';
export const RESERVATION = 'reservation';
export const EDITION_MARKER_BIT_SIZE = 248;

export const findProgramAddress = async (seeds: any, programId: any) => {
  const key = `pda-${seeds.reduce(
    (agg: any, item: any) => agg + item.toString('hex'),
    ''
  )}${programId.toString()}`;
  const cached = localStorage.getItem(key);
  if (cached) {
    const value = JSON.parse(cached);

    return [new PublicKey(value.key), parseInt(value.nonce)];
  }
  const result = await PublicKey.findProgramAddress(seeds, programId);
  localStorage.setItem(
    key,
    JSON.stringify({
      key: result[0].toBase58(),
      nonce: result[1]
    })
  );
  return result;
};

export const decodeMetadata = (buffer: any) => {
  let metadata;
  try {
    metadata = deserializeUnchecked(METADATA_SCHEMA, Metadata, buffer);
  } catch {
    return;
  }
  return metadata;
};

async function getEdition(tokenMint: any) {
  const PROGRAM_IDS = programIds();

  return (
    await findProgramAddress(
      [
        Buffer.from(METADATA_PREFIX),
        PROGRAM_IDS.metadata.toBuffer(),
        tokenMint.toBuffer(),
        Buffer.from(EDITION)
      ],
      PROGRAM_IDS.metadata
    )
  )[0];
}

export class Metadata {
  [x: string]: any;
  constructor(args: any) {
    this.key = MetadataKey.MetadataV1;
    this.updateAuthority = args.updateAuthority;
    this.mint = args.mint;
    this.data = args.data;
    this.primarySaleHappened = args.primarySaleHappened;
    this.isMutable = args.isMutable;
  }

  async init() {
    const edition = await getEdition(this.mint);
    this.edition = edition;
    this.masterEdition = edition;
  }
}

export class MasterEditionV1 {
  [x: string]: number;
  constructor(args: any) {
    this.key = MetadataKey.MasterEditionV1;
    this.supply = args.supply;
    this.maxSupply = args.maxSupply;
    this.printingMint = args.printingMint;
    this.oneTimePrintingAuthorizationMint =
      args.oneTimePrintingAuthorizationMint;
  }
}

export class MasterEditionV2 {
  [x: string]: any;
  constructor(args: any) {
    this.key = MetadataKey.MasterEditionV2;
    this.supply = args.supply;
    this.maxSupply = args.maxSupply;
  }
}

class CreateMetadataArgs {
  [x: string]: any;
  constructor(args: any) {
    this.data = args.data;
    this.isMutable = args.isMutable;
  }
}
class UpdateMetadataArgs {
  [x: string]: any;
  constructor(args: any) {
    this.data = args.data ? args.data : null;
    this.updateAuthority = args.updateAuthority
      ? new PublicKey(args.updateAuthority)
      : null;
    this.primarySaleHappened = args.primarySaleHappened;
  }
}

class CreateMasterEditionArgs {
  [x: string]: any;
  constructor(args: any) {
    this.maxSupply = args.maxSupply;
  }
}

class MintPrintingTokensArgs {
  [x: string]: any;
  constructor(args: any) {
    this.supply = args.supply;
  }
}

export class EditionMarker {
  [x: string]: any;
  constructor(args: any) {
    this.key = MetadataKey.EditionMarker;
    this.ledger = args.ledger;
  }

  editionTaken(edition: any) {
    const editionOffset = edition % EDITION_MARKER_BIT_SIZE;
    const indexOffset = Math.floor(editionOffset / 8);

    if (indexOffset > 30) {
      throw new Error('bad index for edition');
    }

    const positionInBitsetFromRight = 7 - (editionOffset % 8);

    const mask = Math.pow(2, positionInBitsetFromRight);

    const appliedMask = this.ledger[indexOffset] & mask;

    return appliedMask !== 0;
  }
}

export class Edition {
  [x: string]: any;
  constructor(args: any) {
    this.key = MetadataKey.EditionV1;
    this.parent = args.parent;
    this.edition = args.edition;
  }
}
export class Creator {
  [x: string]: any;
  constructor(args: any) {
    this.address = args.address;
    this.verified = args.verified;
    this.share = args.share;
  }
}

export class Data {
  [x: string]: any;
  constructor(args: any) {
    this.name = args.name;
    this.symbol = args.symbol;
    this.uri = args.uri;
    this.sellerFeeBasisPoints = args.sellerFeeBasisPoints;
    this.creators = args.creators;
  }
}

export const METADATA_SCHEMA = new Map([
  [
    CreateMetadataArgs,
    {
      kind: 'struct',
      fields: [
        ['instruction', 'u8'],
        ['data', Data],
        ['isMutable', 'u8'] // bool
      ]
    }
  ],
  [
    UpdateMetadataArgs,
    {
      kind: 'struct',
      fields: [
        ['instruction', 'u8'],
        ['data', { kind: 'option', type: Data }],
        ['updateAuthority', { kind: 'option', type: 'pubkey' }],
        ['primarySaleHappened', { kind: 'option', type: 'u8' }]
      ]
    }
  ],

  [
    CreateMasterEditionArgs,
    {
      kind: 'struct',
      fields: [
        ['instruction', 'u8'],
        ['maxSupply', { kind: 'option', type: 'u64' }]
      ]
    }
  ],
  [
    MintPrintingTokensArgs,
    {
      kind: 'struct',
      fields: [
        ['instruction', 'u8'],
        ['supply', 'u64']
      ]
    }
  ],
  [
    MasterEditionV1,
    {
      kind: 'struct',
      fields: [
        ['key', 'u8'],
        ['supply', 'u64'],
        ['maxSupply', { kind: 'option', type: 'u64' }],
        ['printingMint', 'pubkey'],
        ['oneTimePrintingAuthorizationMint', 'pubkey']
      ]
    }
  ],
  [
    MasterEditionV2,
    {
      kind: 'struct',
      fields: [
        ['key', 'u8'],
        ['supply', 'u64'],
        ['maxSupply', { kind: 'option', type: 'u64' }]
      ]
    }
  ],
  [
    Edition,
    {
      kind: 'struct',
      fields: [
        ['key', 'u8'],
        ['parent', 'pubkey'],
        ['edition', 'u64']
      ]
    }
  ],
  [
    Data,
    {
      kind: 'struct',
      fields: [
        ['name', 'string'],
        ['symbol', 'string'],
        ['uri', 'string'],
        ['sellerFeeBasisPoints', 'u16'],
        ['creators', { kind: 'option', type: [Creator] }]
      ]
    }
  ],
  [
    Creator,
    {
      kind: 'struct',
      fields: [
        ['address', 'pubkey'],
        ['verified', 'u8'],
        ['share', 'u8']
      ]
    }
  ],
  [
    Metadata,
    {
      kind: 'struct',
      fields: [
        ['key', 'u8'],
        ['updateAuthority', 'pubkey'],
        ['mint', 'pubkey'],
        ['data', Data],
        ['primarySaleHappened', 'u8'], // bool
        ['isMutable', 'u8'] // bool
      ]
    }
  ],
  [
    EditionMarker,
    {
      kind: 'struct',
      fields: [
        ['key', 'u8'],
        ['ledger', [31]]
      ]
    }
  ]
]);

export const WRAPPED_SOL_MINT = new PublicKey(
  'So11111111111111111111111111111111111111112'
);
export const TOKEN_PROGRAM_ID = new PublicKey(
  'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA'
);

export const SPL_ASSOCIATED_TOKEN_ACCOUNT_PROGRAM_ID = new PublicKey(
  'ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL'
);
export const BPF_UPGRADE_LOADER_ID = new PublicKey(
  'BPFLoaderUpgradeab1e11111111111111111111111'
);

export const METADATA_PROGRAM_ID = new PublicKey(
  'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s'
  // 'GCUQ7oWCzgtRKnHnuJGxpr5XVeEkxYUXwTKYcqGtxLv4',
);

export const MEMO_ID = new PublicKey(
  'MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr'
);

export const VAULT_ID = new PublicKey(
  'vau1zxA2LbssAUEF7Gpw91zMM1LvXrvpzJtmZ58rPsn'
  // '41cCnZ1Z1upJdtsS1tzFGR34cPFgJLzvJFmgYKpCqkz7',
);

export const AUCTION_ID = new PublicKey(
  'auctxRXPeJoc4817jDhf4HbjnhEcr1cCXenosMhK5R8'
  // '6u5XVthCStUfmNrYhFsST94oKxzwEZfZFHFhiCnB2nR1',
);

export const METAPLEX_ID = new PublicKey(
  'p1exdMJcjVao65QdewkaZRUnU6VPSXhus9n2GzWfh98'
  // '98jcGaKLKx9vv33H9edLUXAydrSipHhJGDQuPXBVPVGp',
);

export const SYSTEM = new PublicKey('11111111111111111111111111111111');

export const ENABLE_FEES_INPUT = false;

// legacy pools are used to show users contributions in those pools to allow for withdrawals of funds
export const PROGRAM_IDS = [
  {
    name: 'mainnet-beta'
  },
  {
    name: 'testnet'
  },

  {
    name: 'devnet'
  },
  {
    name: 'localnet'
  }
];

export const programIds = () => {
  return {
    token: TOKEN_PROGRAM_ID,
    associatedToken: SPL_ASSOCIATED_TOKEN_ACCOUNT_PROGRAM_ID,
    bpf_upgrade_loader: BPF_UPGRADE_LOADER_ID,
    system: SYSTEM,
    metadata: METADATA_PROGRAM_ID,
    memo: MEMO_ID,
    vault: VAULT_ID,
    auction: AUCTION_ID,
    metaplex: METAPLEX_ID
  };
};
